google/field_helpers.go

```diff
@@ -77,6 +82,86 @@ func parseGlobalFieldValue(resourceType, fieldValue, projectSchemaField string,
 	}, nil
 }
 
+type ZonalFieldValue struct {
+	Project string
+	Zone    string
+	Name    string
+
+	resourceType string
+}
+
+func (f ZonalFieldValue) RelativeLink() string {
+	if len(f.Name) == 0 {
+		return ""
+	}
+
+	return fmt.Sprintf(zonalLinkTemplate, f.Project, f.Zone, f.resourceType, f.Name)
+}
+
+// Parses a zonal field supporting 5 different formats:
+// - https://www.googleapis.com/compute/ANY_VERSION/projects/{my_project}/zones/{zone}/{resource_type}/{resource_name}
+// - projects/{my_project}/zones/{zone}/{resource_type}/{resource_name}
+// - zones/{zone}/{resource_type}/{resource_name}
+// - resource_name
+// - "" (empty string). RelativeLink() returns empty if isEmptyValid is true.
+//
+// If the project is not specified, it first tries to get the project from the `projectSchemaField` and then fallback on the default project.
+// If the zone is not specified, it takes the value of `zoneSchemaField`.
+
+func parseZonalFieldValue(resourceType, fieldValue, projectSchemaField, zoneSchemaField string, d TerraformResourceData, config *Config, isEmptyValid bool) (*ZonalFieldValue, error) {
```

Comment:
This function has 7 arguments. Is it doing too much? You might consider breaking it down into smaller functions or somehow grouping related attributes together.  Up to you if you think that's simpler

Reply:
Then entry point would still require all the parameters. The logic using those parameters is extracted away in method like `getProjectFromSchema`.

Reply:
Well, not exactly. Some of these parameters aren't "request-level" (e.g. they don't change inside the specific resource's code) and could be extracted into a more general constructor. You could imagine something like

```go
// Defined at top of resource
var zoneParser := ZoneParser{ResourceType: "networks", IsEmptyValid: false}

func myResourceCreate(d *schema.ResourceData, meta interface{}) {
    // a bunch of stuff happens here
    // [...]

    zoneParser.parse(myFieldValue, "foo_field", "project", d, config)
}
```

Similar to how a regex's pattern is defined globally ahead of where it's actually used. But it's up to you; it may be this representation is uglier.